File: bytebrew-cli/src/infrastructure/lsp/install/BinDirectory.ts
import path from "path";
import fs from "fs/promises";
import fsSync from "fs";
import { ByteBrewHome } from '../../config/ByteBrewHome.js';

const APP_NAME = "vector";

/**
 * Manages the local binary directory for auto-installed LSP servers.
 *
 * Paths per platform:
 *   Linux:   ~/.local/share/vector/bin
 *   macOS:   ~/Library/Application Support/vector/bin
 *   Windows: %APPDATA%/vector/bin
 */
export class BinDirectory {
  private readonly binPath: string;

  constructor() {
    this.binPath = path.join(ByteBrewHome.dataDir(), APP_NAME, "bin");
  }

  getPath(): string {
    return this.binPath;
  }

  async ensureExists(): Promise<void> {
    await fs.mkdir(this.binPath, { recursive: true });
  }

  /**
   * Find a binary in the managed directory.
   * Checks: <binDir>/<name>, <binDir>/node_modules/.bin/<name>
   * On Windows also checks .exe and .cmd extensions.
   */
  async hasBinary(name: string): Promise<string | undefined> {
    const candidates = this.getCandidatePaths(name);
    for (const candidate of candidates) {
      try {
        await fs.access(candidate);
        return candidate;
      } catch {
        // not found, try next
      }
    }
    return undefined;
  }

  /**
   * Synchronous check — used by whichBin() in the hot path.
   * Returns the binary path if found, null otherwise.
   */
  hasBinarySync(name: string): string | null {
    const candidates = this.getCandidatePaths(name);
    for (const candidate of candidates) {
      try {
        fsSync.accessSync(candidate);
        return candidate;
      } catch {
        // not found, try next
      }
    }
    return null;
  }

  /** Get the full path for a binary name (with .exe on Windows) */
  binaryPath(name: string): string {
    const ext = process.platform === "win32" ? ".exe" : "";
    return path.join(this.binPath, name + ext);
  }

  private getCandidatePaths(name: string): string[] {
    const isWin = process.platform === "win32";
    const ext = isWin ? ".exe" : "";
    const candidates: string[] = [
      // Direct binary in bin dir
      path.join(this.binPath, name + ext),
      // npm/bun installed packages
      path.join(this.binPath, "node_modules", ".bin", name + ext),
    ];
    if (isWin) {
      // bun/npm also creates .cmd wrappers on Windows
      candidates.push(
        path.join(this.binPath, "node_modules", ".bin", name + ".cmd"),
      );
    }
    return candidates;
  }
}

File: engine/internal/infrastructure/lsp/bin_directory.go
package lsp

import (
	"os"
	"path/filepath"
	"runtime"
)

// ManagedBinDir returns the path to the managed binary directory for auto-installed LSP servers.
//   - Windows: %APPDATA%/bytebrew/bin/
//   - macOS:   ~/Library/Application Support/bytebrew/bin/
//   - Linux:   ${XDG_DATA_HOME:-~/.local/share}/bytebrew/bin/
func ManagedBinDir() string {
	switch runtime.GOOS {
	case "windows":
		appData := os.Getenv("APPDATA")
		if appData == "" {
			home, _ := os.UserHomeDir()
			appData = filepath.Join(home, "AppData", "Roaming")
		}
		return filepath.Join(appData, "bytebrew", "bin")

	case "darwin":
		home, _ := os.UserHomeDir()
		return filepath.Join(home, "Library", "Application Support", "bytebrew", "bin")

	default: // linux and others
		dataHome := os.Getenv("XDG_DATA_HOME")
		if dataHome == "" {
			home, _ := os.UserHomeDir()
			dataHome = filepath.Join(home, ".local", "share")
		}
		return filepath.Join(dataHome, "bytebrew", "bin")
	}
}

// EnsureBinDir creates the managed binary directory if it doesn't exist.
func EnsureBinDir() (string, error) {
	dir := ManagedBinDir()
	if err := os.MkdirAll(dir, 0755); err != nil {
		return "", err
	}
	return dir, nil
}
